Introduce the west repository to the Zephyr Project (#11715)
zephyrproject-rtos/zephyr: doc/west/index.rst

as a build system generator makes it non-trivial to present the user with
a consistent set of command-line switches and options that enable advanced
usage of the build tool ultimately responsible for building, flashing and
debugging the application.
In order to unify and streamline the developer's experience when working with
Zephyr, west was introduced. It is very important to note that west does not
replace the CMake-based build system, but directly uses it and augments it.

.. _west-struct:

Structure
*********

West is downloaded and installed on a system in two stages:

* Bootstrapper: Installed using ``pip``, implements ``west init``
* Installation: Installed using ``west init``, implements all other commands

Additional information about the two parts can be found below.

Bootstrapper
============

The bootstrapper module is distributed using `PyPi`_ (i.e. :file:`pip`) and it's
placed in the user's ``PATH`` becoming the only entry point to west.
It implements a single command: ``west init``. This command needs to be run
first in order to use the rest of functionality included in ``west``, since
``west init`` will do the following:

> * Clone west itself in a :file:`west/west` folder

Reviewer: I was arguing for having `.west`, like repo has `.repo` (though you don't get building and flashing helpers in there), but people said that users would be horribly confused by the leading dot.

> * Clone west itself in a :file:`west/west` folder

Reviewer 2: > people said that users would be horribly confused by the leading dot.

Actually, people said that we should be consistent on Windows and Linux where file names are concerned.

> * Clone the manifest in a :file:`west/manifest` folder (additional information

Reviewer 3: ```suggestion
* Clone the manifest in a :file:`manifest` folder (additional information
```

  on the manifest can be found in the :ref:`west-multi-repo` section)

Once ``west init`` has been run, the bootstrapper will delegate the handling of
any commands different from ``init`` to the cloned west installation.

This means that there is a single bootstrapper instance installed at any time
(unless you use virtual environments), which can then be used to initialize as
many installations as needed.

.. _west-struct-installation:

Installation
============

> A west installation is the result of running ``west init`` on a folder. As

Reviewer 3: ```suggestion
A west installation is the result of running ``west init`` in an existing folder,
or ``west init <foldername>`` to create and initialize a folder . As
```

described on the previous section, an installation always includes a special
:file:`west/` folder that contains a clone of both west itself and the manifest.
Additionally, and once ``west clone`` has been run, the installation will also
contain a clone of any repositories listed in the manifest. The directory tree
layout of an installation using the default Zephyr manifest therefore looks
like this:

.. code-block:: console

>    ├── west

Reviewer 3: to avoid confusion (and mirror the example in the getting started guide) lets call this top level folder ``zephyrproject``
```suggestion
   ├── zephyrproject
```

   │   ├── manifest
   │   └── west
   └── zephyr

> Additionally, ``west init`` creates a configuration file :file:`west/config`

Reviewer 3: to be clear, is this file ``zephyrproject/west/config`` or ``zephyrproject/config``?  (your picture above had two folders named ``west``)

Author: this is `zephyrproject/west/config`

Author: I've tried to clarify this in my diagram: `west init` creates a `west` folder with 3 items:

* A `west` subfolder
* A `manifest` subfolder
* A `config` file

Reviewer 3: Ah, I see now. (I just tried doing a ``west init``).

that stores configuration metadata about the west installation.

To learn more about west's handling of multiple repositories refer to
:ref:`west-multi-repo`.

The west bootstrapper delegates handling for all commands except ``init`` to

> the installation in :file:`west/west`.

Reviewer 3: ```suggestion
the installation in :file:`zephyrproject/west`.
```

Author: See new diagram once I push

.. _west:

West
####

The Zephyr project includes a swiss-army knife command line tool
named ``west`` (Zephyr is an English name for the Latin
`Zephyrus <https://en.wiktionary.org/wiki/Zephyrus>`_, the ancient Greek god
of the west wind).

West is developed in its own `repository on GitHub`_.

West is used to obtain the source code for the Zephyr project and can also be
used to build, debug, and flash applications.

History and motivation
**********************

West was added to the Zephyr project to fulfill two fundamental requirements:

> * The ability to work with multiple Git repositories

Reviewer 4: I know we talked about this already, but for the benefit of everybody;
Why was `repo` not deemed fit for working with multiple Git repositories?

Author: In short:

* Python 2 only
* Uses symlinks so it doesn't work on windows without an admin command prompt
* Hardcoded gerrit code review
* Poorly documented and maintained

Reviewer 5: I think these repo shortcomings are important to mention somewhere in the documentation itself and not just in this code review, otherwise west-multirepo looks like just another NIH tool.

While adding them drop "poorly documented and maintained" or justify it a bit https://gerrit.googlesource.com/git-repo/
https://gerrit.googlesource.com/git-repo/+log/master

Repo is used by a fair number of projects despite these shortcomings. So you are indirectly saying the world could use a better multirepo tool. I could agree. 
 => Why should [west-]multirepo be specific to Zephyr?

* Avoidance of license incompatibilities between original and imported code
* Reduction in size and scope of the core Zephyr codebase, with additional
  repositories containing optional components instead of being imported
  directly into the tree
* Safety and security certifications
* Enforcement of modularization of the components
* Out-of-tree development based on subsets of the supported boards and SoCs

Command-line interface
======================

Zephyr uses CMake as its build system management tool, as described in
:ref:`application`.
While CMake is a very powerful system and language, its two-stage nature
as a build system generator makes it non-trivial to present the user with
a consistent set of command-line switches and options that enable advanced
usage of the build tool ultimately responsible for building, flashing, and
debugging the application.
West was introduced to unify and streamline the developer's experience when
working with Zephyr. It is very important to note that west does not
replace the CMake-based build system, but directly uses it and augments it.

.. _west-struct:

Structure
*********

West is downloaded and installed on a system in two stages:

> * Bootstrapper: Installed using ``pip``, implements ``west init``

Author: I will let @mbolivar answer this one since he designed the 2-step installation process in the first place, which mirrors Google repo's.

> * Bootstrapper: Installed using ``pip``, implements ``west init``

Reviewer 6: >  which mirrors Google repo's.

Android's repo just needs the repo script to be somewhere to be executed. Not to be installed with `pip`. See for example:
https://github.com/aescolar/ci-dockerfiles/blob/master/shippable/install_bsim.sh#L6 
If I can do that with west, without sideeffects somewhere else in the host, then I'm happy.

> * Bootstrapper: Installed using ``pip``, implements ``west init``

Reviewer 7: RFC from where we can discuss changes to allow `west` bootstrapper to be fetched without `pip`
https://github.com/zephyrproject-rtos/west/pull/133

Once ``west init`` has been run, the bootstrapper will delegate the handling of
any west commands other than ``init`` to the cloned west installation.

This means that there is a single bootstrapper instance installed at any time
(unless you use virtual environments), which can then be used to initialize as
many installations as needed.

.. _west-struct-installation:

Installation
============

A west installation is the result of running ``west init``  in an existing
folder, or ``west init <foldername>`` to create and initialize a folder. As
described on the previous section, an installation always includes a special
:file:`west/` folder that contains a clone of both west itself and the manifest.
Additionally, and once ``west clone`` has been run, the installation will also
contain a clone of any repositories listed in the manifest. The directory tree
layout of an installation using the default Zephyr manifest therefore looks
like this:

.. code-block:: console

   └── zephyrproject
       ├── west
       │   ├── config
       │   ├── manifest
       │   └── west
       └── zephyr

Additionally, ``west init`` creates a configuration file :file:`west/config`

> that stores configuration metadata about the west installation.

Reviewer 7: no, this config file is about `west` and `manifest` repo themselves.

The Zephyr project is described in the manifest file.

West is:

- **Optional**: it is always *possible* to drop back to "raw"
  command-line tools, i.e. use Zephyr without using west (although west itself

>   might need to be installed and accessible to the build system). It may not

Reviewer 7: meaning, the tools inside Zephyr, that was already `west`, such as `flashing, debug, attach, ...`

Those are now in `west` repo, so in order to have those working, you would need the `west` repo.
But if not using `west` for flashing, debug, etc. I assume you can leave it out, although we might need to ensure CMake can handle it, in case `west` is not present.

>   might need to be installed and accessible to the build system). It may not

Reviewer 6: > I assume you can leave it out

That being the point

>   might need to be installed and accessible to the build system). It may not

Author: It depends what you do really. If you are not going to flash or debug, then you definitely not need it. In the future however, you will need it in order to build, since the build system will use `west` to locate the different repositories available for building.

>   might need to be installed and accessible to the build system). It may not

Reviewer 7: question is, if people really want / has a reason to avoid cloning `west`.
And Zephyr needs it for tje build system to locate all required repositories for building, then could we provide an easy to do so way, e.g. arguments to CMake. to allow people to do so.

Simply, not preventing them from doing so.
That is give the option that what `west` returns can simply be provided manually ?

>   might need to be installed and accessible to the build system). It may not

Reviewer 6: > In the future however, you will need it in order to build

Ok.. so today you can still build images with that warning, but not doc. And soon nothing. Meaning, not optional :-p

There are many valid usecases to avoid using a meta-tool such as west which has
been custom-designed for a particular project and its requirements.
This section applies to Zephyr users who fall into one or more of these
categories:

- Already use a multi-repository management system (for example Google repo)

- Do not need additional functionality beyond compiling and linking

In order to obtain the Zephyr source code and build it without the help of
west you will need to manually clone the repositories listed in the
`manifest file`_ one by one, at the path indicated in it.

.. code-block:: console

   mkdir zephyrproject
   cd zephyrproject
   git clone https://github.com/zephyrproject-rtos/zephyr
   # clone additional repositories listed in the manifest file

If you want to manage Zephyr's repositories without west but still need to
use west's additional functionality (flashing, debugging, etc.) it is possible
to do so by manually creating an installation:

.. code-block:: console

   # cd into zephyrproject if not already there
   mkdir west
   cd west
   git clone https://github.com/zephyrproject-rtos/west.git
   git clone https://github.com/zephyrproject-rtos/manifest.git

>    echo > .west_topdir

Reviewer 6: `touch .west_topdir` doesn't do it?
